test/integration/storageversion/gc_test.go

```diff
@@ -0,0 +1,235 @@
+
+	kubeclient, err := kubernetes.NewForConfig(result.ClientConfig)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	informers := informers.NewSharedInformerFactory(kubeclient, time.Second)
+	leaseInformer := informers.Coordination().V1().Leases()
+	storageVersionInformer := informers.Internal().V1alpha1().StorageVersions()
+
+	controller := storageversiongc.NewStorageVersionGC(kubeclient, leaseInformer, storageVersionInformer)
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	go leaseInformer.Informer().Run(stopCh)
+	go storageVersionInformer.Informer().Run(stopCh)
+	go controller.Run(stopCh)
+
+	createTestAPIServerIdentityLease(t, kubeclient, idA)
+	createTestAPIServerIdentityLease(t, kubeclient, idB)
+
+	t.Run("storage version with non-existing id should be GC'ed", func(t *testing.T) {
+		createTestStorageVersion(t, kubeclient, idNonExist)
+		assertStorageVersionDeleted(t, kubeclient)
+	})
+
+	t.Run("storage version with valid id should not be GC'ed", func(t *testing.T) {
+		createTestStorageVersion(t, kubeclient, idA)
+		time.Sleep(10 * time.Second)
+		sv, err := kubeclient.InternalV1alpha1().StorageVersions().Get(
+			context.TODO(), svName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to retrieve valid storage version: %v", err)
+		}
+		if len(sv.Status.StorageVersions) != 1 {
+			t.Errorf("unexpected number of storage version entries, expected 1, got: %v",
+				sv.Status.StorageVersions)
+		}
+		expectedID := idA
+		if sv.Status.StorageVersions[0].APIServerID != expectedID {
+			t.Errorf("unexpected storage version entry id, expected %v, got: %v",
+				expectedID, sv.Status.StorageVersions[0].APIServerID)
+		}
+		assertCommonEncodingVersion(t, kubeclient, pointer.StringPtr("v0"))
+		if err := kubeclient.InternalV1alpha1().StorageVersions().Delete(
+			context.TODO(), svName, metav1.DeleteOptions{}); err != nil {
+			t.Fatalf("failed to cleanup valid storage version: %v", err)
+		}
+	})
+
+	t.Run("deleting an id should delete a storage version entry that it owns", func(t *testing.T) {
+		createTestStorageVersion(t, kubeclient, idA, idB)
+		assertStorageVersionEntries(t, kubeclient, 2, idA)
+		assertCommonEncodingVersion(t, kubeclient, nil)
+		deleteTestAPIServerIdentityLease(t, kubeclient, idA)
+		assertStorageVersionEntries(t, kubeclient, 1, idB)
+		assertCommonEncodingVersion(t, kubeclient, pointer.StringPtr("v1"))
+	})
+
+	t.Run("deleting an id should delete a storage version object that it owns entirely", func(t *testing.T) {
+		deleteTestAPIServerIdentityLease(t, kubeclient, idB)
+		assertStorageVersionDeleted(t, kubeclient)
+	})
+}
+
+func createTestStorageVersion(t *testing.T, client kubernetes.Interface, ids ...string) {
+	sv := &apiserverinternalv1alpha1.StorageVersion{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: svName,
+		},
+	}
+	for i, id := range ids {
+		v := apiserverinternalv1alpha1.ServerStorageVersion{
+			APIServerID:       id,
+			EncodingVersion:   fmt.Sprintf("v%d", i),
```

Comment:
This is fragile. The encoding version depends on the order of the ids.

How about encodingVersion = fmt.Sprintf("v%s", strings.TrimPrefix(id, "id"))? Also validating the ids are prefixed with "id" at the beginning of this function, and otherwise t.Fatalf(...)

Reply:
that's better. Done

I'll rewrite it later to just have a `map[id string]v1alpha1.ServerStorageVersion` somewhere